src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/ResettableCompletionSource.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information
+
+using System.Threading.Tasks;
+using System.Threading.Tasks.Sources;
+
+namespace System.Net.Quic.Implementations.MsQuic.Internal
+{
+    /// <summary>
+    /// A resettable completion source which can be completed multiple times.
+    /// Used to make methods async between completed events and their associated async method.
+    /// </summary>
+    internal abstract class ResettableCompletionSource<T> : IValueTaskSource<T>
+    {
+        protected ManualResetValueTaskSourceCore<T> _valueTaskSource;
+
+        public ResettableCompletionSource()
+        {
+            _valueTaskSource.RunContinuationsAsynchronously = true;
```

Comment:
:/

Reply:
wut.

Reply:
There's a perf hit here. Are we trying to avoid doing large work on the msquic callback thread?

Reply:
Yeah, we need to get off of the thread to return to MsQuic ASAP. We do the same in our Libuv implementation.

Reply:
Yes. You can't want to do a lot of work on the msquic thread, as it will delay processing the network protocol.

Reply:
Sounds good.